generate xla codegen in-tree

bdhirsh opened the pull request:
This PR ports the codegen logic from the [XLA repo](https://github.com/pytorch/xla/blob/master/scripts/gen.py) to be in-tree.

This particular PR maintains (almost) byte-for-byte compatibility with XLA's codegen, with the differences highlighted below. That means that there's some xla-specific stuff in the codegen, which will be removed in subsequent PR's.

**Plan**
[] (this PR) merge byte-for-byte version of XLA codegen in-tree
[] Get XLA to use the in-tree codegen
[] Remove the xla-specific `bridge::` API from the codegen
[] Kill the CPU fallback codegen in favor of a boxed kernel
[] Remove other xla-specific bits, like the file names (`aten_xla_type.h`)
[] update codegen to properly handle generating `inplace`/`out`
[] update codegen to handle structured

**New codegen workflow**
I have a corresponding PR in the `pytorch/xla` repo that uses these changes [here](https://github.com/pytorch/xla/pull/2869).
After checking out pytorch/xla from source, you can run `python ../tools/codegen/gen_backend_stubs.py  --output_dir="torch_xla/csrc/" --source_yaml="xla_native_functions.yaml"`. It spits 3 files, which are close-to-byte-for-byte-identical to what pytorch/xla currently has:
* `torch_xla/csrc/aten_xla_type.h`
* `torch_xla/csrc/aten_xla_type_default.h`
* `torch_xla/csrc/aten_xla_type_default.cpp`

`aten_xla_type.h` is currently maintained manually in the pytorch/xla repo; instead, a new `xla_native_functions.yaml` file will maintain the list of ops that xla supports, and the codegen will generated the header file. This is more similar to what our other in-tree codegen does. Unfortunately I couldn't keep it byte-for-byte identically because the original order of the signatures was arbitrary, but the file contains the same set of declarations with the same signatures.


**Byte-for-byte changes**

`aten_xla_type_default.h` contains the headers for functions that fall back to CPU, same as before (identical sans comments and whitespace, and `Stream` -> `at::Stream` [they had a `using c10::Stream` somewhere]).

`aten_xla_type_default.cpp` contains a few things:
* kernels that fall back to CPU for any non-composite ops that xla doesn't support
* code-generated out kernels for a few operators, which just call into the xla-implemented kernel
* dispatcher registrations for all relevant kernels (xla kernels, codegen'd out wrappers, codegen'd fallbacks)

That file is byte-for-byte identical with a few differences:
* namespaces: `c10::Device` ->`at::Device`, and `Stream` -> `at::Stream`
* The other diffs are smaller and easy to look over, I pasted them here: P408712288
  * We no longer manually unpack `TensorOptions` in the CPU fallbacks, and instead use the faithful C++ API
  * The codegen'd CPU fallback for `miopen_rnn_backward` had a small bug that this codegen fixes: it wasn't converting all output tensor arguments from CPU back to XLA properly.


**Data model changes**
* New `ExternalBackendMetadata`, `ExternalBackendFunction`, and `ExternalBackendFunctionsGroup` classes. `ExternalBackendMetadata` represents a line from `xla_native_functions.yaml`, and `ExternalBackendFunction/ExternalBackendFunctionsGroup` are data representations that link the metadata to the corresponding native functions.


**NOTE: What is staying/leaving after this PR**
The bulk of the changes live in `gen_external_fallbacks.py`. It contains:
* declarations/definitions for CPU fallbacks
* code-generated `out` wrappers (same as what XLA does)
* all dispatcher registrations

All of the CPU fallback logic will hopefully be killed in the near future, when I write I replace it with a boxed fallback. I opted to put it in a new file (`gen_external_fallbacks.py`) because (a) most of it will disappear soon, making it easier to disentangle later, and (b) it made merge conflicts (slightly) easier to deal with while working on this PR :)

When that logic is killed, I'm planning to move the remaining logic (`codegen'd out wrappers + dispatcher registrations) into `register_dispatch_key.py`, since the logic will eventually look a lot more similar.



Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55796 remove bridge API from codegen
* #55795 add _to_cpu() operator
* **#55050 generate xla codegen in-tree**

Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

bdhirsh committed:
[EXPERIMENTAL, WIP] generate xla codegen in-tree

[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet




[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet




[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet




[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet




[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet


Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet


Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet


Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet


Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet


Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet


Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet


Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet


Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet


Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet


Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet


Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

not ready for review yet


Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh committed:
Update on "[EXPERIMENTAL, WIP] generate xla codegen in-tree"

This PR ports the codegen logic from the [XLA repo](https://github.com/pytorch/xla/blob/master/scripts/gen.py) to be in-tree.

This particular PR maintains (almost) byte-for-byte compatibility with XLA's codegen, with the differences highlighted below. That means that there's some xla-specific stuff in the codegen, which will be removed in subsequent PR's.

**Plan**
[] (this PR) merge byte-for-byte version of XLA codegen in-tree
[] Get XLA to use the in-tree codegen
[] Remove the xla-specific `bridge::` API from the codegen
[] Kill the CPU fallback codegen in favor of a boxed kernel
[] Remove other xla-specific bits, like the file names (`aten_xla_type.h`)
[] update codegen to properly handle generating `inplace`/`out`
[] update codegen to handle structured

**New codegen workflow**
I have a corresponding PR in the `pytorch/xla` repo that uses these changes [here](https://github.com/pytorch/xla/pull/2869).
After checking out pytorch/xla from source, you can run `python ../tools/codegen/gen_backend_stubs.py  --output_dir="torch_xla/csrc/" --source_yaml="xla_native_functions.yaml"`. It spits 3 files, which are close-to-byte-for-byte-identical to what pytorch/xla currently has:
* `torch_xla/csrc/aten_xla_type.h`
* `torch_xla/csrc/aten_xla_type_default.h`
* `torch_xla/csrc/aten_xla_type_default.cpp`

`aten_xla_type.h` is currently maintained manually in the pytorch/xla repo; instead, a new `xla_native_functions.yaml` file will maintain the list of ops that xla supports, and the codegen will generated the header file. This is more similar to what our other in-tree codegen does. Unfortunately I couldn't keep it byte-for-byte identically because the original order of the signatures was arbitrary, but the file contains the same set of declarations with the same signatures.


**Byte-for-byte changes**

`aten_xla_type_default.h` contains the headers for functions that fall back to CPU, same as before (identical sans comments and whitespace, and `Stream` -> `at::Stream` [they had a `using c10::Stream` somewhere]).

`aten_xla_type_default.cpp` contains a few things:
* kernels that fall back to CPU for any non-composite ops that xla doesn't support
* code-generated out kernels for a few operators, which just call into the xla-implemented kernel
* dispatcher registrations for all relevant kernels (xla kernels, codegen'd out wrappers, codegen'd fallbacks)

That file is byte-for-byte identical with a few differences:
* namespaces: `c10::Device` ->`at::Device`, and `Stream` -> `at::Stream`
* The other diffs are smaller and easy to look over, I pasted them here: P408712288
  * We no longer manually unpack `TensorOptions` in the CPU fallbacks, and instead use the faithful C++ API
  * The codegen'd CPU fallback for `miopen_rnn_backward` had a small bug that this codegen fixes: it wasn't converting all output tensor arguments from CPU back to XLA properly.


**Data model changes**
* New `ExternalBackendMetadata`, `ExternalBackendFunction`, and `ExternalBackendFunctionsGroup` classes. `ExternalBackendMetadata` represents a line from `xla_native_functions.yaml`, and `ExternalBackendFunction/ExternalBackendFunctionsGroup` are data representations that link the metadata to the corresponding native functions.


**NOTE: What is staying/leaving after this PR**
The bulk of the changes live in `gen_external_fallbacks.py`. It contains:
* declarations/definitions for CPU fallbacks
* code-generated `out` wrappers (same as what XLA does)
* all dispatcher registrations

All of the CPU fallback logic will hopefully be killed in the near future, when I write I replace it with a boxed fallback. I opted to put it in a new file (`gen_external_fallbacks.py`) because (a) most of it will disappear soon, making it easier to disentangle later, and (b) it made merge conflicts (slightly) easier to deal with while working on this PR :)

When that logic is killed, I'm planning to move the remaining logic (`codegen'd out wrappers + dispatcher registrations) into `register_dispatch_key.py`, since the logic will eventually look a lot more similar.




Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh commented on tools/codegen/api/types.py:
```diff
@@ -266,11 +266,16 @@ def no_default(self) -> 'Binding':
-        return f"{self.type} {self.name}{mb_default}"
+
+        # casting only needs to know the type
+        if func_ptr_cast:
```
sigh... we actually only need this for CPU fallbacks, so I'll add a note to kill it when we make a boxed kernel.

bdhirsh commented on tools/codegen/api/dispatcher.py:
```diff
@@ -40,27 +41,25 @@ def returns_type(rs: Sequence[Return]) -> CType:
-        return argument(a.dtype) + argument(a.layout) + argument(a.device) + argument(a.pin_memory)
-    else:
-        assert_never(a)
+def jit_arguments(func: FunctionSchema) -> List[Argument]:
```
mostly re-jiggered this so that the xla codegen could pattern match on the Argument types, and have them match 1-to-1 with the arguments in each xla kernel (which follow the dispatcher API)

bhosmer replied:
just curious, why the "jit" in "jit_arguments"?

bdhirsh replied:
Mostly
(1) This function returns a list of `Argument` objects which model the jit schema; the only aspect of this function that's dispatcher-specific is that it unpacks `SelfArgument` and `TensorOptionsArguments`.
(2) `def arguments` is already taken and I can't overload in python 😛 if you have a better name in mind feel free.

ezyang commented on tools/codegen/api/cpp.py:
```diff
@@ -167,11 +171,14 @@ def return_names(f: NativeFunction) -> Sequence[str]:
             else:
                 name = r.name
-        # If there is no explicit name, we just name the output result,
+        # If there is no explicit name and no overide was passed in, we just name the output result,
```
nit: override

ezyang commented on tools/codegen/api/cpp.py:
```diff
@@ -146,7 +150,7 @@ def returns_type(rs: Sequence[Return]) -> CType:
         return TupleCType([return_type(r) for r in rs])
 
-def return_names(f: NativeFunction) -> Sequence[str]:
+def return_names(f: NativeFunction, *, override_name: Optional[str] = None) -> Sequence[str]:
```
override_name doesn't seem all that appropriate here as half the time you don't actually override when this is non-None! Maybe `fallback_name="result"` (and then you can also dump "result") below.

ezyang commented on tools/codegen/api/types.py:
```diff
@@ -266,11 +266,16 @@ def no_default(self) -> 'Binding':
         )
 
-    def decl(self) -> str:
+    def decl(self, *, func_ptr_cast: bool = False) -> str:
```
So... what does `func_ptr_cast` do anyway?

ezyang replied:
Why isn't this just not a separate method on its own? (Or even just self.type?)

bdhirsh replied:
yeah it totally should be

bdhirsh replied:
oh, `DispatcherSignature` already has a `type` function that does everything I need except making the type a pointer. I'll add a flag there.

E.g.
`at::Tensor(const at::Tensor &, const at::Tensor &, const at::Scalar &)`
vs
`at::Tensor (*)(const at::Tensor &, const at::Tensor &, const at::Scalar &)`

ezyang replied:
oh yeah, that's kind of unfortunate. I don't think there's any way to do it compositionally without breaking out using

```
using f = int(int);
using fptr = f*;
```

but that's inconvenient for codegen, so this will have to do.

bhosmer commented on tools/codegen/context.py:
```diff
@@ -8,19 +9,33 @@
     else:
         f = g
-    with context(f'in {f.loc}:\n  {f.func}'):
+    with context(f'in native_functions.yaml line {f.loc}:\n  {f.func}'):
```
I'm not sure there's any simple way to make the file name less hardcoded here, but it would be nice. E.g. if file loading in gen.py and this here both referred to a global or something. Not a huge deal but would forestall drift later etc.

bhosmer commented on tools/codegen/dest/gen_external_aten_fallbacks.py:
```diff
@@ -0,0 +1,305 @@
+
+    out_name = 'x_result'
+    if tuple_idx is not None:
+        out_name = f"std::get<{tuple_idx}>(x_result)"
```
The `x_result` embedded here should probably be `{out_name}`... more importantly though, you probably want to hoist `x_result` up out of the various string literals it's embedded in and turn it into a global, since it's a shared point of definition for the codegen.

bdhirsh replied:
agreed.

Actually, I don't think it needs to be a global; it's used by both the cpu fallback and the codegen'd `out` wrappers, but there's no reason that the variable name needs to be the same between the two aside from byte-for-byte compatibility. The important thing is just that it's defined in one place / consistent everywhere within the CPU fallback, and everywhere within the `out` wrapper.

So lmk if you have a different opinion, but I'm leaning towards defining it twice, once in the `out`-wrapper codegen and once in the CPU fallback codegen. That'll also make things less messy later, when I move the `out` wrapper codegen into a different file from the fallback.

bdhirsh replied:
oh scratch that, I just double checked and `x_result` isn't actually needed in the `out` wrapper codegen. It's only used in the CPU fallback.

bhosmer reviewed (commented):
Hey LG! Took a first look and left a few initial comments, back later for more 😁

bdhirsh commented on aten/src/ATen/templates/aten_xla_type.h:
```diff
@@ -0,0 +1,22 @@
+  static void InitializeAtenBindings();
+
+  //////////////////////////////////////////////////////////////////////////////
+  // ATEN API ovverrides in alphabetical order.
```
dunno what it is with my brain, second PR this week that I mis-spelled overrides 😛

ezyang commented on aten/src/ATen/templates/aten_xla_type.h:
```diff
@@ -0,0 +1,22 @@
+#pragma once
+// Autogenerated file by gen_backend_stubs.py. Do not edit directly!
```
use `// ${generated_comment}` instead

ezyang commented on tools/codegen/dest/native_functions.py:
```diff
@@ -22,12 +25,21 @@ def gen_unstructured(f: NativeFunction) -> List[str]:
-        returns_type = native.returns_type(f.func.returns).cpp_type()
-        args = native.arguments(f.func)
-        rs.append(f"TORCH_API {returns_type} {n}({', '.join(a.decl() for a in args)});")
+        rs.append(f"TORCH_API {native_sig.decl(name=n)};")
```
This refactor seems like it could have been done in a prep commit. Maybe not that long so not worth it.

bdhirsh replied:
yep it's totally unrelated. I noticed it and made the change after I'd already made a bunch of changes to this file, so I thought it would be more effort than it was worth to split up. But that might be saying something about my workflow more than anything else.

ezyang commented on tools/codegen/gen.py:
```diff
@@ -553,11 +553,11 @@ def compute_cpp_argument_yaml(cpp_a: Binding, *, schema_order: bool, kwarg_only_
-def compute_argument_yaml(a: Argument, *, schema_order: bool, kwarg_only_set: Set[str],
-                          out_arg_set: Set[str], name_to_field_name: Dict[str, str]) -> object:
+def compute_argument_yaml(a: Argument, *, schema_order: bool, kwarg_only_set: Set[str], out_arg_set: Set[str],
+                          name_to_field_name: Dict[str, str]) -> object:
```
Did anything happen here?

bdhirsh replied:
Nope. I think I previously added an argument to these functions and later removed it, but forgot to revert the formatting changes that I made.

It's confusing in this PR, so I'm just going to fix it

ezyang commented on tools/codegen/model.py:
```diff
@@ -1335,6 +1335,97 @@ def __str__(self) -> str:
             return f"{self.name}"
 
+@dataclass(frozen=True)
+class ExternalBackendMetadata:
```
An alternative data model, which would require less structs to be defined here, is to index the external backend metadata separately from ExternalBackendFunction, rather than insisting it be directly contained here. Then, anywhere you previously would have taken an ExternalBackendFunction, you instead take a NativeFunction and then look up the ExternalBackendMetadata from the index.

bdhirsh replied:
Yeah, that reduces the amount every data class stuff that goes in here, at the cost of requiring all of the code that acts on ExternalBackendFunctions to do a little more work.

I have a slight preference for the beefed-up data model. The properties I defined on `ExternalBackendFunction` save a fair amount of redundant work from all of the callers, and will abstract any future logic we add to the metadata; e.g. the fact that not having a metadata -> the backend hasn't implemented that op, or that `primary` will figure out if the backend favors `out` vs `functional` operators. But let me know if you feel strongly one way or the other.

ezyang commented on tools/codegen/gen_backend_stubs.py:
```diff
@@ -0,0 +1,114 @@
+import pathlib
+import argparse
+import os
+import yaml
```
For ease of in tree development, we're probably going to want to setup some tests that exercise this code path on simplified YAML files that cover important functionality.

ezyang replied:
In particular, unlike in tree native_functions.yaml, we are going to need testing for error cases and making sure we give appropriate errors in those cases. Expect tests can help make this painless.

ezyang commented on tools/codegen/gen_backend_stubs.py:
```diff
@@ -0,0 +1,114 @@
+
+    backend = yaml_values['backend']
+    supported = yaml_values['supported']
+    supported_autograd = yaml_values['autograd']
```
As this YAML is user provided it is worth being more paranoid about the inputs provided here. Good input validation here will help save you from weird error messages later. `NativeFunction.from_yaml` does a pretty good job of validating its arguments and giving useful error information when things have gone wrong, and I recommend modeling the code here after it, if you don't have ideas on your own.

bdhirsh replied:
You're right, thanks. I'm going to make the error'ing a little more robust in this PR. I'll probably add some explicit tests in a followup.

ezyang commented on tools/codegen/gen_backend_stubs.py:
```diff
@@ -0,0 +1,114 @@
+    template_dir = os.path.join(pytorch_root, "aten/src/ATen/templates")
+
+    def make_file_manager(install_dir: str) -> FileManager:
+        return FileManager(install_dir=install_dir, template_dir=template_dir, dry_run=False)
```
NB: file manager is pointless unless you let people `dry_run` it. The abstraction solely exists to let people dry run the codegen script so that they can find out what files it writes out.

bhosmer replied:
I thought FileManager (in non-dry-run mode) also suppressed output if it was unchanged? (Not that supporting dry_run here is a bad idea)

bdhirsh replied:
ah, got it. It doesn't seem unreasonable to allow an optional `dry_run` flag to the external codegen, similar to what `gen.py` has.

ezyang commented on tools/codegen/dest/gen_external_aten_fallbacks.py:
```diff
@@ -0,0 +1,305 @@
+    @method_with_native_function
+    def __call__(self, g: Union[ExternalBackendFunctionsGroup, ExternalBackendFunction]) -> List[str]:
+
+        def gen_out_wrapper(g: ExternalBackendFunctionsGroup) -> Optional[str]:
```
Why not make these methods on GenExternalAtenFallback? Would save you a level of indentation, and brings this file more in line with the prevailing style of dest modules.

bdhirsh replied:
One reservation that I had was for generating out wrappers: given an `ExternalBackendFunctionsGroup`, `gen_unstructured_external()` can't really operate on each `ExternalBackendFunction` independently: the codegen is supposed to generate an `out` wrapper if the `out` ExternalBackendFunction doesn't have ExternalMetadata, and that logic needs to know the name of the `functional` op (in order to call it). When the functions are nested as closures like they are now, `gen_unstructured_external()` has access to the higher-scope `ExternalBackendFunctionsGroup`.

Locally I ended up just doing what you said, but I had to update the signature to `gen_unstructured_external(f: ExternalBackendFunction, g: ExternalBackendFunctionGroup)`, so it has all of the information it needs. I think it's an overall win for readability, since that level of indentation made things pretty confusing.

bdhirsh replied:
(I'm actually not gonna make this change for now, and just land the PR as is. lmk if you think like the change though)

ezyang replied:
Well, what you said here doesn't seem to apply to `gen_out_wrapper`. And it is totally reasonable to pass it in as an extra argument, that's what we do in other places.

I'm not *too* allergic to nested closures, but I know @bhosmer has been on my case about them in the past, and explicitly lifting out the closures does make it easier to reason about what information is propagated and what is not.

bdhirsh replied:
Removing the nested closures in this file will also be that much easier after I finish moving the `gen_out_wrapper` logic out of this file as part of https://github.com/pytorch/pytorch/pull/56835. I can do that in a follow-up.

ezyang commented on tools/codegen/dest/gen_external_aten_fallbacks.py:
```diff
@@ -0,0 +1,305 @@
+
+            def get_device_param(args: List[Argument]) -> str:
+                # TODO: the XLA codegen has specific precedence rules when determining which tensor argument
+                # to use as the device argument.
```
cc @wenleix

wenleix replied:
for the XLA case, do they expect all tensors are on the same device? :)

bdhirsh replied:
Good question, I dunno :) maybe @ailzhang does.

We probably want to keep the logic for choosing which input to get the device from consistent either way though, right?

bdhirsh replied:
It looks like mixed xla/cpu devices are allowed, at least:
```
>>> import torch_xla.core.xla_model as xm
>>> dev = xm.xla_device()
>>> a = torch.ones(3, device=dev)
>>> b = torch.ones(3)
>>> a + b
tensor([2., 2., 2.], device='xla:0')
>>> a.device
device(type='xla', index=0)
>>> b.device
device(type='cpu')
>>>
```

ezyang commented on tools/codegen/dest/gen_external_aten_fallbacks.py:
```diff
@@ -0,0 +1,305 @@
+                    namespace = 'AtenXlaTypeDefault'
+                payload = f"static_cast<{dispatcher_sig.decl(func_ptr_cast=True)}>(&{namespace}::{name})"
+                return f'  m.impl("{f.native_function.func.name}", {payload});\n'
+
```
assert something about target here?

ezyang commented:
Substantively this PR looks pretty good. What would you like to do? Land as is and start merging fixups, or keep working on this until you're closer?

bdhirsh commented on tools/codegen/context.py:
```diff
@@ -8,19 +9,33 @@
+    ExternalBackendFunctionsGroup,
+    Union[NativeFunction, NativeFunctionsGroup],
+    Union[ExternalBackendFunctionsGroup, ExternalBackendFunction],
+    Union[NativeFunction, NativeFunctionsGroup, ExternalBackendFunction, ExternalBackendFunctionsGroup]
```
sigh :'(

dunno how to make mypy happy without an exponential blowup

bdhirsh commented:
> Substantively this PR looks pretty good. What would you like to do? Land as is and start merging fixups, or keep working on this until you're closer?

I'm hoping to land this PR sooner rather than later, but a bunch of the feedback looks like small things that will be easier to fix directly in the PR. I'm leaning towards making most of the quick fixes first.

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -754,6 +754,26 @@ def get_custom_build_selector(
+            return [r]
+
+    # TODO: how come ValuesView isn't a Sequence lol
+    return list(concatMap(flatten_pre_group, list(pre_grouped_native_functions.values())))
```
nope, identical. Just moved it into a function so `gen_backend_stubs` can re-use it. I guess the diff tool had trouble seeing that 😛

bdhirsh commented on tools/codegen/model.py:
```diff
@@ -1335,6 +1335,97 @@ def __str__(self) -> str:
+        if kind == SchemaKind.out or kind == SchemaKind.inplace:
+            assert self.metadata is None or not self.metadata.structured, \
+                "Found an out/inplace operator marked with the structured keyword." \
+                f" Only functional operators can be marked as structured. operator={str(self.native_function.func.name)}"
```
yep, should have add that before

bdhirsh committed:
Update on "generate xla codegen in-tree"

This PR ports the codegen logic from the [XLA repo](https://github.com/pytorch/xla/blob/master/scripts/gen.py) to be in-tree.

This particular PR maintains (almost) byte-for-byte compatibility with XLA's codegen, with the differences highlighted below. That means that there's some xla-specific stuff in the codegen, which will be removed in subsequent PR's.

**Plan**
[] (this PR) merge byte-for-byte version of XLA codegen in-tree
[] Get XLA to use the in-tree codegen
[] Remove the xla-specific `bridge::` API from the codegen
[] Kill the CPU fallback codegen in favor of a boxed kernel
[] Remove other xla-specific bits, like the file names (`aten_xla_type.h`)
[] update codegen to properly handle generating `inplace`/`out`
[] update codegen to handle structured

**New codegen workflow**
I have a corresponding PR in the `pytorch/xla` repo that uses these changes [here](https://github.com/pytorch/xla/pull/2869).
After checking out pytorch/xla from source, you can run `python ../tools/codegen/gen_backend_stubs.py  --output_dir="torch_xla/csrc/" --source_yaml="xla_native_functions.yaml"`. It spits 3 files, which are close-to-byte-for-byte-identical to what pytorch/xla currently has:
* `torch_xla/csrc/aten_xla_type.h`
* `torch_xla/csrc/aten_xla_type_default.h`
* `torch_xla/csrc/aten_xla_type_default.cpp`

`aten_xla_type.h` is currently maintained manually in the pytorch/xla repo; instead, a new `xla_native_functions.yaml` file will maintain the list of ops that xla supports, and the codegen will generated the header file. This is more similar to what our other in-tree codegen does. Unfortunately I couldn't keep it byte-for-byte identically because the original order of the signatures was arbitrary, but the file contains the same set of declarations with the same signatures.


**Byte-for-byte changes**

`aten_xla_type_default.h` contains the headers for functions that fall back to CPU, same as before (identical sans comments and whitespace, and `Stream` -> `at::Stream` [they had a `using c10::Stream` somewhere]).

`aten_xla_type_default.cpp` contains a few things:
* kernels that fall back to CPU for any non-composite ops that xla doesn't support
* code-generated out kernels for a few operators, which just call into the xla-implemented kernel
* dispatcher registrations for all relevant kernels (xla kernels, codegen'd out wrappers, codegen'd fallbacks)

That file is byte-for-byte identical with a few differences:
* namespaces: `c10::Device` ->`at::Device`, and `Stream` -> `at::Stream`
* The other diffs are smaller and easy to look over, I pasted them here: P408712288
  * We no longer manually unpack `TensorOptions` in the CPU fallbacks, and instead use the faithful C++ API
  * The codegen'd CPU fallback for `miopen_rnn_backward` had a small bug that this codegen fixes: it wasn't converting all output tensor arguments from CPU back to XLA properly.


**Data model changes**
* New `ExternalBackendMetadata`, `ExternalBackendFunction`, and `ExternalBackendFunctionsGroup` classes. `ExternalBackendMetadata` represents a line from `xla_native_functions.yaml`, and `ExternalBackendFunction/ExternalBackendFunctionsGroup` are data representations that link the metadata to the corresponding native functions.


**NOTE: What is staying/leaving after this PR**
The bulk of the changes live in `gen_external_fallbacks.py`. It contains:
* declarations/definitions for CPU fallbacks
* code-generated `out` wrappers (same as what XLA does)
* all dispatcher registrations

All of the CPU fallback logic will hopefully be killed in the near future, when I write I replace it with a boxed fallback. I opted to put it in a new file (`gen_external_fallbacks.py`) because (a) most of it will disappear soon, making it easier to disentangle later, and (b) it made merge conflicts (slightly) easier to deal with while working on this PR :)

When that logic is killed, I'm planning to move the remaining logic (`codegen'd out wrappers + dispatcher registrations) into `register_dispatch_key.py`, since the logic will eventually look a lot more similar.




Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]

bdhirsh committed:
Update on "generate xla codegen in-tree"

This PR ports the codegen logic from the [XLA repo](https://github.com/pytorch/xla/blob/master/scripts/gen.py) to be in-tree.

This particular PR maintains (almost) byte-for-byte compatibility with XLA's codegen, with the differences highlighted below. That means that there's some xla-specific stuff in the codegen, which will be removed in subsequent PR's.

**Plan**
[] (this PR) merge byte-for-byte version of XLA codegen in-tree
[] Get XLA to use the in-tree codegen
[] Remove the xla-specific `bridge::` API from the codegen
[] Kill the CPU fallback codegen in favor of a boxed kernel
[] Remove other xla-specific bits, like the file names (`aten_xla_type.h`)
[] update codegen to properly handle generating `inplace`/`out`
[] update codegen to handle structured

**New codegen workflow**
I have a corresponding PR in the `pytorch/xla` repo that uses these changes [here](https://github.com/pytorch/xla/pull/2869).
After checking out pytorch/xla from source, you can run `python ../tools/codegen/gen_backend_stubs.py  --output_dir="torch_xla/csrc/" --source_yaml="xla_native_functions.yaml"`. It spits 3 files, which are close-to-byte-for-byte-identical to what pytorch/xla currently has:
* `torch_xla/csrc/aten_xla_type.h`
* `torch_xla/csrc/aten_xla_type_default.h`
* `torch_xla/csrc/aten_xla_type_default.cpp`

`aten_xla_type.h` is currently maintained manually in the pytorch/xla repo; instead, a new `xla_native_functions.yaml` file will maintain the list of ops that xla supports, and the codegen will generated the header file. This is more similar to what our other in-tree codegen does. Unfortunately I couldn't keep it byte-for-byte identically because the original order of the signatures was arbitrary, but the file contains the same set of declarations with the same signatures.


**Byte-for-byte changes**

`aten_xla_type_default.h` contains the headers for functions that fall back to CPU, same as before (identical sans comments and whitespace, and `Stream` -> `at::Stream` [they had a `using c10::Stream` somewhere]).

`aten_xla_type_default.cpp` contains a few things:
* kernels that fall back to CPU for any non-composite ops that xla doesn't support
* code-generated out kernels for a few operators, which just call into the xla-implemented kernel
* dispatcher registrations for all relevant kernels (xla kernels, codegen'd out wrappers, codegen'd fallbacks)

That file is byte-for-byte identical with a few differences:
* namespaces: `c10::Device` ->`at::Device`, and `Stream` -> `at::Stream`
* The other diffs are smaller and easy to look over, I pasted them here: P408712288
  * We no longer manually unpack `TensorOptions` in the CPU fallbacks, and instead use the faithful C++ API
  * The codegen'd CPU fallback for `miopen_rnn_backward` had a small bug that this codegen fixes: it wasn't converting all output tensor arguments from CPU back to XLA properly.


**Data model changes**
* New `ExternalBackendMetadata`, `ExternalBackendFunction`, and `ExternalBackendFunctionsGroup` classes. `ExternalBackendMetadata` represents a line from `xla_native_functions.yaml`, and `ExternalBackendFunction/ExternalBackendFunctionsGroup` are data representations that link the metadata to the corresponding native functions.


**NOTE: What is staying/leaving after this PR**
The bulk of the changes live in `gen_external_fallbacks.py`. It contains:
* declarations/definitions for CPU fallbacks
* code-generated `out` wrappers (same as what XLA does)
* all dispatcher registrations

All of the CPU fallback logic will hopefully be killed in the near future, when I write I replace it with a boxed fallback. I opted to put it in a new file (`gen_external_fallbacks.py`) because (a) most of it will disappear soon, making it easier to disentangle later, and (b) it made merge conflicts (slightly) easier to deal with while working on this PR :)

When that logic is killed, I'm planning to move the remaining logic (`codegen'd out wrappers + dispatcher registrations) into `register_dispatch_key.py`, since the logic will eventually look a lot more similar.




Differential Revision: [D27708346](https://our.internmc.facebook.com/intern/diff/D27708346)

[ghstack-poisoned]
